Objects/complexobject.c

```diff
@@ -474,83 +579,87 @@ complex_hash(PyComplexObject *v)
-    return PyComplex_FromCComplex(result);
+    pc->imag = 0.0;
+    return real_to_double(pobj, &(pc->real));
 }
 
-static PyObject *
-complex_sub(PyObject *v, PyObject *w)
-{
-    Py_complex result;
-    Py_complex a, b;
-    TO_COMPLEX(v, a);
-    TO_COMPLEX(w, b);
-    result = _Py_c_diff(a, b);
-    return PyComplex_FromCComplex(result);
-}
+/* Complex arithmetic rules implement special mixed-mode case where combining
+   a pure-real (float or int) value and a complex value is performed directly
+   without first coercing the real value to a complex value.
 
-static PyObject *
-complex_mul(PyObject *v, PyObject *w)
-{
-    Py_complex result;
-    Py_complex a, b;
-    TO_COMPLEX(v, a);
-    TO_COMPLEX(w, b);
-    result = _Py_c_prod(a, b);
-    return PyComplex_FromCComplex(result);
-}
+   Let us consider the addition as an example, assuming that ints are implicitly
+   converted to floats.  We have the following rules (up to variants with changed
+   order of operands):
 
-static PyObject *
-complex_div(PyObject *v, PyObject *w)
-{
-    Py_complex quot;
-    Py_complex a, b;
-    TO_COMPLEX(v, a);
-    TO_COMPLEX(w, b);
-    errno = 0;
-    quot = _Py_c_quot(a, b);
-    if (errno == EDOM) {
-        PyErr_SetString(PyExc_ZeroDivisionError, "division by zero");
-        return NULL;
-    }
-    return PyComplex_FromCComplex(quot);
-}
+       complex(x, y) + complex(u, v) = complex(x + u, y + v)
+       float(x) + complex(u, v) = complex(x + u, v)
+
+   Similar rules are implemented for subtraction, multiplication and division.
+   See C11's Annex G, sections G.5.1 and G.5.2.
+ */
+
+#define COMPLEX_BINOP(name, func)                             \
+    static PyObject *                                         \
+    complex_##name(PyObject *v, PyObject *w)                  \
+    {                                                         \
+        Py_complex a;                                         \
+        errno = 0;                                            \
+        if (PyComplex_Check(w)) {                             \
+            Py_complex b = ((PyComplexObject *)(w))->cval;    \
+            if (PyComplex_Check(v)) {                         \
+                a = ((PyComplexObject *)(v))->cval;           \
+                a = _Py_c_##func(a, b);                       \
+            }                                                 \
+            else if (real_to_double(&v, &a.real) < 0) {       \
+                return v;                                     \
+            }                                                 \
+            else {                                            \
+                a = _Py_rc_##func(a.real, b);                 \
+            }                                                 \
+        }                                                     \
+        else {                                                \
+            a = ((PyComplexObject *)(v))->cval;               \
```

Comment:
I just observed this but don't you actually need to check that `v` is a complex as well? Because you assume that you either have (complex, complex) or (complex, something else). So I would add an assertion on `v` for this path (because you are calling `_Py_cr_##func` afterwards)

Reply:
> I just observed this but don't you actually need to check that v is a complex as well?

One argument must be complex (or it's subtype).  But you are right, probably it's possible to call functions somehow ignoring usual protocols for binary operations.

Other classes (e.g.) do such check, so I will.  Thanks.